File: ios/AIEnglish/ReadingAnalysisEngine.swift
import Foundation

struct ReadingAnalysis {
    var mode: String
    var paragraphGists: [(index: Int, gistZh: String, keySentenceEn: String)]
    var coreViewpointZh: String
    var examPoints: [String]
    var logicRelations: [String]
    var howToSolveZh: String
}

enum ReadingAnalysisEngine {
    private static let staticHowToSolve = """
1. 先读题干与选项，圈出题干关键词（人名、地名、数字、大写、段落号）。
2. 细节题用关键词回原文定位，同义改写多为正确选项。
3. 主旨/标题题重点读首段、尾段与各段首句，排除片面细节项。
4. 词义猜测题看该词前后句的同义复述、举例或对比线索。
5. 推断题只选「能从原文合理推出」的选项，忌主观臆断与过度引申。
6. 注意路标词：转折、因果、递进、举例，常对应考点与正确项。
"""

    private static let logicPatterns: [(NSRegularExpression, String)] = {
        let pairs: [(String, String)] = [
            (#"\b(however|but|yet|instead|while)\b"#, "转折对比"),
            (#"\b(because|since|as|so that)\b"#, "原因目的"),
            (#"\b(therefore|thus|so|as a result|consequently)\b"#, "结果结论"),
            (#"\b(although|though|even if)\b"#, "让步"),
            (#"\b(for example|for instance|such as)\b"#, "举例说明"),
            (#"\b(first|second|finally|in conclusion|in short)\b"#, "顺序与总结"),
            (#"\b(if|unless|when)\b"#, "条件时间"),
            (#"\b(and|also|besides)\b"#, "并列递进"),
        ]
        return pairs.compactMap { pair in
            (try? NSRegularExpression(pattern: pair.0, options: .caseInsensitive)).map { ($0, pair.1) }
        }
    }()

    static func analyzeHeuristic(text: String) -> ReadingAnalysis? {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard t.count >= 15 else { return nil }
        let paras = splitParagraphs(t)
        let gists: [(Int, String, String)] = paras.enumerated().map { i, p in
            (i + 1, "（启发式）以下为首句/关键句参考，完整段意请结合全段核对；印刷或 OCR 可能有误差。", firstSentence(p))
        }
        var logic = detectLogic(t)
        if logic.isEmpty {
            logic = ["（启发式）未检出明显路标词时，仍留意 and / but / because 等基础逻辑连接词。"]
        }
        return ReadingAnalysis(
            mode: "heuristic",
            paragraphGists: gists,
            coreViewpointZh: "（未使用 AI）请通读各段首句、末段与反复出现的主题词，用一句话概括「作者最想传递的信息」；说明文常为「对象 + 特征/步骤」，议论文关注论点与论据。",
            examPoints: detectExamHints(t),
            logicRelations: logic,
            howToSolveZh: staticHowToSolve
        )
    }

    private static func splitParagraphs(_ text: String) -> [String] {
        var parts = text.components(separatedBy: "\n\n").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
        if parts.count <= 1, text.count > 380 {
            let sentenceRe = try! NSRegularExpression(pattern: #"[^.!?]+[.!?]+"#, options: [])
            let range = NSRange(text.startIndex..., in: text)
            let matches = sentenceRe.matches(in: text, options: [], range: range)
            let sentences = matches.compactMap { m -> String? in
                guard let r = Range(m.range, in: text) else { return nil }
                return String(text[r])
            }
            if !sentences.isEmpty {
                var chunks: [String] = []
                var i = 0
                while i < sentences.count {
                    let slice = sentences[i..<min(i + 3, sentences.count)]
                    chunks.append(slice.joined(separator: " "))
                    i += 3
                }
                if chunks.count > 1 { parts = chunks }
            }
        }
        return parts.isEmpty ? [text] : parts
    }

    private static func firstSentence(_ p: String) -> String {
        let s = p.trimmingCharacters(in: .whitespacesAndNewlines)
        if let r = s.range(of: #"^[\s\S]{1,400}?[.!?](?=\s|$)"#, options: .regularExpression) {
            return String(s[r])
        }
        let line = s.split(separator: "\n", omittingEmptySubsequences: false).first.map(String.init) ?? s
        return String(line.prefix(220))
    }

    private static func detectLogic(_ text: String) -> [String] {
        var found = Set<String>()
        for (re, zh) in logicPatterns {
            let range = NSRange(text.startIndex..., in: text)
            if re.firstMatch(in: text, options: [], range: range) != nil {
                found.insert(zh)
            }
        }
        return found.map { "文中可体现「\($0)」类逻辑衔接，做题时注意选项是否与该逻辑一致。" }
    }

    private static func detectExamHints(_ text: String) -> [String] {
        var out: [String] = []
        let rules: [(String, String)] = [
            (#"\b(according to the passage|the passage says|paragraph\s*\d)\b"#, "信息定位题：用题干关键词回原文找依据，排除未提及或偷换概念的选项。"),
            (#"\b(main(ly)?\s+idea|best title|purpose of|mainly about)\b"#, "主旨/标题类：串联各段中心句，勿把某一例子或细节当作全文中心。"),
            (#"\b(infer|imply|suggest|it can be learned)\b"#, "推断题：选项必须是原文可合理推出的结论，忌绝对化与无中生有。"),
            (#"\b(the word|refers to|closest in meaning)\b"#, "词义/指代题：看前后句同义复述、举例或反义对比；指代向前找最近名词。"),
            (#"\b(order|sequence|correct order)\b"#, "排序题：抓住时间词与 first/then/finally 等路标，先定位再排顺序。"),
        ]
        for (pat, msg) in rules {
            if (try? NSRegularExpression(pattern: pat, options: .caseInsensitive))?.firstMatch(
                in: text, options: [], range: NSRange(text.startIndex..., in: text)
            ) != nil {
                out.append(msg)
            }
        }
        let extra = [
            "长难句：先找主谓宾，再看从句与插入语，避免被生词干扰整体理解。",
            "对比题：注意 while、whereas、unlike 等引出的对照关系。",
        ]
        for e in extra where out.count < 8 { out.append(e) }
        return Array(Set(out)).prefix(8).map { $0 }
    }

    /// 解析 `AIService.analyzeReading` 返回的 JSON 字符串（可含 markdown 围栏）。
    static func parseAIResponse(_ raw: String) -> ReadingAnalysis? {
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasPrefix("```") {
            if let firstNl = s.firstIndex(of: "\n") {
                s = String(s[s.index(after: firstNl)...]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
            if let range = s.range(of: "```", options: .backwards) {
                s = String(s[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        guard let data = s.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        let pg = obj["paragraphGists"] as? [[String: Any]] ?? []
        var gists: [(Int, String, String)] = []
        for item in pg {
            let idx = item["index"] as? Int ?? (gists.count + 1)
            let g = (item["gistZh"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let k = (item["keySentenceEn"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            gists.append((idx, g, k))
        }
        let core = (obj["coreViewpointZh"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let exam = (obj["examPoints"] as? [Any])?.compactMap { $0 as? String } ?? []
        let logic = (obj["logicRelations"] as? [Any])?.compactMap { $0 as? String } ?? []
        let how = (obj["howToSolveZh"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if gists.isEmpty && core.isEmpty && exam.isEmpty && logic.isEmpty && how.isEmpty { return nil }
        return ReadingAnalysis(
            mode: "ai",
            paragraphGists: gists,
            coreViewpointZh: core,
            examPoints: exam,
            logicRelations: logic,
            howToSolveZh: how
        )
    }
}

File: android/app/src/main/java/com/cltf/aienglish/domain/ReadingAnalysisEngine.kt
package com.cltf.aienglish.domain

import org.json.JSONArray
import org.json.JSONObject
import java.util.regex.Pattern

object ReadingAnalysisEngine {

    private val staticHowToSolve = """
        |1. 先读题干与选项，圈出题干关键词（人名、地名、数字、大写、段落号）。
        |2. 细节题用关键词回原文定位，同义改写多为正确选项。
        |3. 主旨/标题题重点读首段、尾段与各段首句，排除片面细节项。
        |4. 词义猜测题看该词前后句的同义复述、举例或对比线索。
        |5. 推断题只选「能从原文合理推出」的选项，忌主观臆断与过度引申。
        |6. 注意路标词：转折、因果、递进、举例，常对应考点与正确项。
    """.trimMargin()

    private val logicPatterns: List<Pair<Pattern, String>> = listOf(
        Pattern.compile("\\b(however|but|yet|instead|while)\\b", Pattern.CASE_INSENSITIVE) to "转折对比",
        Pattern.compile("\\b(because|since|as|so that)\\b", Pattern.CASE_INSENSITIVE) to "原因目的",
        Pattern.compile("\\b(therefore|thus|so|as a result|consequently)\\b", Pattern.CASE_INSENSITIVE) to "结果结论",
        Pattern.compile("\\b(although|though|even if)\\b", Pattern.CASE_INSENSITIVE) to "让步",
        Pattern.compile("\\b(for example|for instance|such as)\\b", Pattern.CASE_INSENSITIVE) to "举例说明",
        Pattern.compile("\\b(first|second|finally|in conclusion|in short)\\b", Pattern.CASE_INSENSITIVE) to "顺序与总结",
        Pattern.compile("\\b(if|unless|when)\\b", Pattern.CASE_INSENSITIVE) to "条件时间",
        Pattern.compile("\\b(and|also|besides)\\b", Pattern.CASE_INSENSITIVE) to "并列递进",
    )

    fun analyzeHeuristic(text: String): ReadingAnalysis? {
        val t = text.trim()
        if (t.length < 15) return null
        val paras = splitParagraphs(t)
        val gists = paras.mapIndexed { i, p ->
            ParagraphGist(
                index = i + 1,
                gistZh = "（启发式）以下为首句/关键句参考，完整段意请结合全段核对；印刷或 OCR 可能有误差。",
                keySentenceEn = firstSentence(p)
            )
        }
        var logic = detectLogic(t)
        if (logic.isEmpty()) {
            logic = listOf("（启发式）未检出明显路标词时，仍留意 and / but / because 等基础逻辑连接词。")
        }
        return ReadingAnalysis(
            mode = "heuristic",
            paragraphGists = gists,
            coreViewpointZh = "（未使用 AI）请通读各段首句、末段与反复出现的主题词，用一句话概括「作者最想传递的信息」；说明文常为「对象 + 特征/步骤」，议论文关注论点与论据。",
            examPoints = detectExamHints(t),
            logicRelations = logic,
            howToSolveZh = staticHowToSolve
        )
    }

    fun parseAIResponse(raw: String): ReadingAnalysis? {
        var s = raw.trim()
        if (s.startsWith("```")) {
            val firstNl = s.indexOf('\n')
            if (firstNl >= 0) s = s.substring(firstNl + 1).trim()
            val idx = s.lastIndexOf("```")
            if (idx >= 0) s = s.substring(0, idx).trim()
        }
        return runCatching {
            val obj = JSONObject(s)
            val pg = obj.optJSONArray("paragraphGists") ?: JSONArray()
            val gists = buildList {
                for (i in 0 until pg.length()) {
                    val item = pg.optJSONObject(i) ?: continue
                    add(
                        ParagraphGist(
                            index = item.optInt("index", size + 1),
                            gistZh = item.optString("gistZh", "").trim(),
                            keySentenceEn = item.optString("keySentenceEn", "").trim()
                        )
                    )
                }
            }
            val core = obj.optString("coreViewpointZh", "").trim()
            val exam = obj.optJSONArray("examPoints")?.toStringList() ?: emptyList()
            val logic = obj.optJSONArray("logicRelations")?.toStringList() ?: emptyList()
            val how = obj.optString("howToSolveZh", "").trim()
            if (gists.isEmpty() && core.isEmpty() && exam.isEmpty() && logic.isEmpty() && how.isEmpty()) return@runCatching null
            ReadingAnalysis(
                mode = "ai",
                paragraphGists = gists,
                coreViewpointZh = core,
                examPoints = exam,
                logicRelations = logic,
                howToSolveZh = how
            )
        }.getOrNull()
    }

    private fun JSONArray.toStringList(): List<String> = buildList {
        for (i in 0 until length()) {
            val v = opt(i)
            if (v is String && v.isNotBlank()) add(v.trim())
        }
    }

    private fun splitParagraphs(text: String): List<String> {
        var parts = text.split("\n\n").map { it.trim() }.filter { it.isNotEmpty() }
        if (parts.size <= 1 && text.length > 380) {
            val sentenceRe = Pattern.compile("[^.!?]+[.!?]+")
            val m = sentenceRe.matcher(text)
            val sentences = mutableListOf<String>()
            while (m.find()) sentences.add(m.group())
            if (sentences.isNotEmpty()) {
                val chunks = mutableListOf<String>()
                var i = 0
                while (i < sentences.size) {
                    val end = minOf(i + 3, sentences.size)
                    chunks.add(sentences.subList(i, end).joinToString(" "))
                    i += 3
                }
                if (chunks.size > 1) parts = chunks
            }
        }
        return if (parts.isEmpty()) listOf(text) else parts
    }

    private fun firstSentence(p: String): String {
        val s = p.trim()
        val re = Pattern.compile("^[\\s\\S]{1,400}?[.!?](?=\\s|$)")
        val m = re.matcher(s)
        if (m.find()) return m.group()
        val line = s.lineSequence().firstOrNull() ?: s
        return line.take(220)
    }

    private fun detectLogic(text: String): List<String> {
        val found = linkedSetOf<String>()
        for ((re, zh) in logicPatterns) {
            if (re.matcher(text).find()) found.add(zh)
        }
        return found.map { "文中可体现「$it」类逻辑衔接，做题时注意选项是否与该逻辑一致。" }
    }

    private fun detectExamHints(text: String): List<String> {
        val rules = listOf(
            Pattern.compile("\\b(according to the passage|the passage says|paragraph\\s*\\d)\\b", Pattern.CASE_INSENSITIVE) to
                "信息定位题：用题干关键词回原文找依据，排除未提及或偷换概念的选项。",
            Pattern.compile("\\b(main(ly)?\\s+idea|best title|purpose of|mainly about)\\b", Pattern.CASE_INSENSITIVE) to
                "主旨/标题类：串联各段中心句，勿把某一例子或细节当作全文中心。",
            Pattern.compile("\\b(infer|imply|suggest|it can be learned)\\b", Pattern.CASE_INSENSITIVE) to
                "推断题：选项必须是原文可合理推出的结论，忌绝对化与无中生有。",
            Pattern.compile("\\b(the word|refers to|closest in meaning)\\b", Pattern.CASE_INSENSITIVE) to
                "词义/指代题：看前后句同义复述、举例或反义对比；指代向前找最近名词。",
            Pattern.compile("\\b(order|sequence|correct order)\\b", Pattern.CASE_INSENSITIVE) to
                "排序题：抓住时间词与 first/then/finally 等路标，先定位再排顺序。"
        )
        val out = mutableListOf<String>()
        for ((pat, msg) in rules) {
            if (pat.matcher(text).find()) out.add(msg)
        }
        val extra = listOf(
            "长难句：先找主谓宾，再看从句与插入语，避免被生词干扰整体理解。",
            "对比题：注意 while、whereas、unlike 等引出的对照关系。"
        )
        for (e in extra) {
            if (out.size < 8) out.add(e)
        }
        return out.distinct().take(8)
    }
}
